aioxmpp/bookmarks/service.py

```diff
@@ -290,58 +303,107 @@ def sync(self):
         yield from self.get_bookmarks()
 
     @asyncio.coroutine
-    def add_bookmark(self, new_bookmark):
+    def add_bookmark(self, new_bookmark, max_retries=3):
         """
-        Add a bookmark.
+        Add a bookmark and check whether it was successfully added to the
+        bookmark list. Already existant bookmarks are not added twice.
 
-        Already existant bookmarks are not added twice.
+        :raises RuntimeError: if the bookmark is not in the bookmark list
+                              after `max_retries` retries.
         """
-        bookmarks = yield from self._get_bookmarks()
-        for bookmark in bookmarks:
-            if bookmark == new_bookmark:
-                break
-        else:
-            bookmarks.append(new_bookmark)
-        yield from self._set_bookmarks(bookmarks)
+        with (yield from self._lock):
```

Comment:
The worst that could happen is outdated state, that will be fixed by the next call to sync. But agreed, it could happen, that we our state does not correspond to the latest received bookmark state.